controllers/harbor/harbor.go

```diff
@@ -30,18 +36,82 @@ func (harbor *HarborReconciler) Reconcile() (*lcm.CRStatus, error) {
 		if errors.IsNotFound(err) {
 			return harbor.Provision()
 		} else {
-			// TODO given clear reason and message.
-			return harborCRNotReadyStatus("", ""), err
+			return harborClusterCRNotReadyStatus("", ""), err
 		}
 	}
-	return nil, nil
+	harbor.CurrentHarborCR = &harborCR
+	event := harbor.checkReconcileEvent(harbor.HarborCluster, &harborCR)
+	switch event {
+	case ScalingEvent:
+		return harbor.Scale()
+	case UpdatingEvent:
+		return harbor.Update(harbor.HarborCluster)
+	}
+
+	err = harbor.Get(harbor.getHarborCRNamespacedName(), &harborCR)
+	if err != nil {
+		return harborClusterCRUnknownStatus(), err
+	}
+	return harborClusterCRStatus(&harborCR), nil
 }
 
-func (harbor *HarborReconciler) Delete() (*lcm.CRStatus, error) {
-	panic("implement me")
+// unsetReplicas will set replicas to nil for all components in v1alpha1.Harbor.
+// This is a helper method to check whether harbor cr is equal expect replicas.
+func unsetReplicas(harbor *v1alpha1.Harbor) {
+	if harbor.Spec.Components.Core != nil {
```

Comment:
The definitions of Harbor itself including:

```go
type HarborSpec struct {
  // ...
  Components HarborComponents `json:"components,omitempty"`
  // ...
}

type HarborComponents struct {
	// +optional
	Core *CoreComponent `json:"core,omitempty"`

	// +optional
	Portal *PortalComponent `json:"portal,omitempty"`

	// +optional
	Registry *RegistryComponent `json:"registry,omitempty"`

	// +optional
	JobService *JobServiceComponent `json:"jobService,omitempty"`

	// +optional
	ChartMuseum *ChartMuseumComponent `json:"chartMuseum,omitempty"`

	// +optional
	Clair *ClairComponent `json:"clair,omitempty"`

	// +optional
	Notary *NotaryComponent `json:"notary,omitempty"`
}

// Use core component as an example

type CoreComponent struct {
	HarborDeployment `json:",inline"`

	// +kubebuilder:validation:Required
	DatabaseSecret string `json:"databaseSecret"`
}

// Every components inherited from  HarborDeployment
type HarborDeployment struct {
	// Number of desired pods. This is a pointer to distinguish between explicit
	// zero and not specified. Defaults to 1.
	// +optional
	// +kubebuilder:validation:Minimum=1
	Replicas *int32 `json:"replicas,omitempty"`

	// +optional
	Image *string `json:"image,omitempty"`

	// +optional
	NodeSelector     NodeSelector                  `json:"nodeSelector,omitempty"`
	ImagePullSecrets []corev1.LocalObjectReference `json:"imagePullSecrets,omitempty" patchStrategy:"merge" patchMergeKey:"name"`
}
```

So, can we use `HarborDeployment` loop to set the common properties such as `Replicas `

```go
func resetReplicas(spec *HarborSpec) {
  for _, comp := range spec.Componnets {
      if hComp, ok := comp.(HarborDeployment); ok {
             hComp.Replicas = nil
     }
  }
}

func setReplica(spec *HarborSpec, expectedR int) {
   // similar logic with above reset func
}
```

Reply:
It need to transform HarborComponents to []HarborDeployment